RecoTauTag/RecoTau/src/AntiElectronDeadECAL.cc

```diff
@@ -0,0 +1,170 @@
+#include "CondFormats/EcalObjects/interface/EcalChannelStatus.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+#include "Geometry/CaloGeometry/interface/CaloGeometry.h"
+#include "Geometry/CaloTopology/interface/EcalTrigTowerConstituentsMap.h"
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "DataFormats/TauReco/interface/PFTau.h"
+#include "DataFormats/PatCandidates/interface/Tau.h"
+#include "DataFormats/DetId/interface/DetId.h"
+#include "DataFormats/EcalDetId/interface/EBDetId.h"
+#include "DataFormats/EcalDetId/interface/EEDetId.h"
+#include "DataFormats/Math/interface/deltaR.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include <TMath.h>
+
+AntiElectronDeadECAL::AntiElectronDeadECAL(const edm::ParameterSet& cfg) : isFirstEvent_(true) {
+  minStatus_ = cfg.getParameter<uint32_t>("minStatus");
+  dR_ = cfg.getParameter<double>("dR");
+
+  extrapolateToECalEntrance_ = cfg.getParameter<bool>("extrapolateToECalEntrance");
+
+  verbosity_ = cfg.getParameter<int>("verbosity");
+}
+
+AntiElectronDeadECAL::~AntiElectronDeadECAL() {}
+
+void AntiElectronDeadECAL::beginEvent(const edm::EventSetup& es) {
+  updateBadTowers(es);
+  positionAtECalEntrance_.beginEvent(es);
+}
+
+namespace {
+  template <class Id>
+  void loopXtals(std::map<uint32_t, unsigned>& nBadCrystals,
+                 std::map<uint32_t, unsigned>& maxStatus,
+                 std::map<uint32_t, double>& sumEta,
+                 std::map<uint32_t, double>& sumPhi,
+                 const EcalChannelStatus* channelStatus,
+                 const CaloGeometry* caloGeometry,
+                 const EcalTrigTowerConstituentsMap* ttMap,
+                 unsigned minStatus,
+                 const uint16_t statusMask) {
+    // NOTE: modified version of SUSY CAF code
+    //         UserCode/SusyCAF/plugins/SusyCAF_EcalDeadChannels.cc
+    for (int i = 0; i < Id::kSizeForDenseIndexing; ++i) {
+      Id id = Id::unhashIndex(i);
+      if (id == Id(0))
+        continue;
+      EcalChannelStatusMap::const_iterator it = channelStatus->getMap().find(id.rawId());
+      unsigned status = (it == channelStatus->end()) ? 0 : (it->getStatusCode() & statusMask);
+      if (status >= minStatus) {
+        const GlobalPoint& point = caloGeometry->getPosition(id);
+        uint32_t key = ttMap->towerOf(id);
+        maxStatus[key] = TMath::Max(status, maxStatus[key]);
+        ++nBadCrystals[key];
+        sumEta[key] += point.eta();
+        sumPhi[key] += point.phi();
+      }
+    }
+  }
+}  // namespace
+
+void AntiElectronDeadECAL::updateBadTowers(const edm::EventSetup& es) {
+  // NOTE: modified version of SUSY CAF code
+  //         UserCode/SusyCAF/plugins/SusyCAF_EcalDeadChannels.cc
+  const uint32_t channelStatusId = es.get<EcalChannelStatusRcd>().cacheIdentifier();
+  const uint32_t caloGeometryId = es.get<CaloGeometryRecord>().cacheIdentifier();
+  const uint32_t idealGeometryId = es.get<IdealGeometryRecord>().cacheIdentifier();
```

Comment:
Done, but it will be welcome that you check this implementation. My test passed successfully, but I am not sure indeed conditions were uploaded when changed.